File: backend/app/execution/capabilities/aging.py
"""aging — bucket dates into aging categories (0-30, 31-60, 61-90, 90+)."""

from __future__ import annotations

import re
from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    date_column = params.get("dateColumn")
    output_column = params.get("outputColumn")
    buckets = params.get("buckets") or [30, 60, 90]
    reference_date = params.get("referenceDate")
    has_headers = bool(params.get("hasHeaders", True))

    if not date_column or not output_column:
        return {"status": "error", "message": "aging requires 'dateColumn' and 'outputColumn'."}

    date_raw = resolve_range(ctx.workbook_handle, date_column)
    date_used = date_raw.current_region if date_raw.count > 1 else date_raw

    addr = date_used.address
    addr_tail = addr.split("!")[-1]
    m = re.match(r"^\$?([A-Z]+)\$?(\d+):\$?([A-Z]+)\$?(\d+)$", addr_tail)
    if not m:
        # Single-cell / single-column fallback.
        m2 = re.match(r"^\$?([A-Z]+)\$?(\d+)$", addr_tail)
        if not m2:
            return {"status": "error", "message": f"Could not parse date column address: {addr}"}
        col_l = m2.group(1)
        top_row = int(m2.group(2))
        bot_row = top_row
    else:
        col_l = m.group(1)
        top_row = int(m.group(2))
        bot_row = int(m.group(4))

    first_data_row = top_row + 1 if has_headers else top_row
    last_data_row = bot_row
    data_count = max(0, last_data_row - first_data_row + 1)
    if data_count == 0:
        return {"status": "success", "message": "No data rows in dateColumn."}

    sheet_name = date_used.sheet.name
    sheet_prefix = f"'{sheet_name}'!" if " " in sheet_name else f"{sheet_name}!"

    # Normalize output column.
    out_col = output_column
    out_sheet_name = None
    if "!" in out_col:
        parts = out_col.split("!", 1)
        out_sheet_name = parts[0].strip("'")
        out_col = parts[1]
    out_col = re.sub(r"[:$]", "", out_col)
    out_col = re.sub(r"\d+$", "", out_col).upper()
    if not re.match(r"^[A-Z]+$", out_col):
        return {"status": "error", "message": f"Invalid outputColumn: {output_column!r}"}

    out_sheet = ctx.workbook_handle.sheets[out_sheet_name] if out_sheet_name else date_used.sheet

    # Reference date expression.
    if reference_date:
        s = str(reference_date).strip()
        rm = re.match(r"^(\d{1,2})/(\d{1,2})/(\d{2,4})$", s)
        if rm:
            y = int(rm.group(3))
            if y < 100:
                y += 2000
            ref_expr = f"DATE({y},{int(rm.group(2))},{int(rm.group(1))})"
        else:
            iso = re.match(r"^(\d{4})-(\d{1,2})-(\d{1,2})$", s)
            if iso:
                ref_expr = f"DATE({int(iso.group(1))},{int(iso.group(2))},{int(iso.group(3))})"
            else:
                ref_expr = f'DATEVALUE("{s}")'
    else:
        ref_expr = "TODAY()"

    sorted_b = sorted(buckets)
    labels: list[str] = []
    for i, upper in enumerate(sorted_b):
        lower = 0 if i == 0 else sorted_b[i - 1] + 1
        labels.append(f"{lower}-{upper}")
    labels.append(f"{sorted_b[-1]}+")

    first_date_cell = f"{col_l}{first_data_row}"
    age_expr = f"({ref_expr}-{first_date_cell})"
    ifs_parts = [f'{age_expr}<={sorted_b[i]},"{labels[i]}"' for i in range(len(sorted_b))]
    ifs_parts.append(f'TRUE,"{labels[-1]}"')
    formula = f"=IFS({','.join(ifs_parts)})"

    if ctx.dry_run:
        return {"status": "preview", "message": f"Would write aging buckets for {data_count} row(s)."}

    try:
        if has_headers:
            out_sheet.range(f"{out_col}{top_row}").value = "Age Bucket"
        # Write formula to every data row. xlwings handles relative ref
        # adjustment when formulas are supplied as a list (same formula per
        # row, Excel adjusts the source cell row).
        target = out_sheet.range(f"{out_col}{first_data_row}:{out_col}{last_data_row}")
        target.formula = [[formula]] * data_count
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"Write failed: {exc}", "error": str(exc)}

    output_addr = f"{sheet_prefix}{out_col}{first_data_row}:{out_col}{last_data_row}"
    return {
        "status": "success",
        "message": f"Wrote aging buckets ({', '.join(labels)}) to {output_addr}.",
        "outputs": {"outputRange": output_addr},
    }


registry.register("aging", handler, mutates=True)

File: frontend/src/engine/capabilities/aging.ts
/**
 * aging — bucket dates into aging categories.
 *
 * Given a date column and a list of upper-bound day counts (default
 * [30, 60, 90]), emits a formula column that labels each row with the
 * matching bucket. The open-ended last bucket reads `{lastBucket}+`.
 *
 * Example: dateColumn = "Sheet1!C2:C100", buckets = [30, 60, 90]
 *   age(date) ≤ 30   → "0-30"
 *   30 < age ≤ 60    → "31-60"
 *   60 < age ≤ 90    → "61-90"
 *   age > 90         → "90+"
 */

import { CapabilityMeta, AgingParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "aging",
  description: "Bucket dates into aging categories (0-30, 31-60, 61-90, 90+) with a formula column",
  mutates: true,
  affectsFormatting: false,
  requiresApiSet: "ExcelApi 1.2",
};

async function handler(
  context: Excel.RequestContext,
  params: AgingParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { dateColumn, outputColumn, buckets = [30, 60, 90], referenceDate, hasHeaders = true } = params;

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would write aging buckets (${buckets.join(",")}+) for ${dateColumn} into ${outputColumn}.`,
    };
  }

  // Resolve the date column so we know where data starts/ends.
  const dateRaw = resolveRange(context, dateColumn);
  const dateUsed = dateRaw.getUsedRange(false);
  dateUsed.load(["address", "rowCount", "worksheet/name"]);
  await context.sync();

  const addrPart = dateUsed.address.includes("!") ? dateUsed.address.split("!").pop()! : dateUsed.address;
  const m = addrPart.match(/^\$?([A-Z]+)\$?(\d+):\$?([A-Z]+)\$?(\d+)$/);
  if (!m) {
    return { stepId: "", status: "error", message: `Could not parse date column address: ${dateUsed.address}` };
  }
  const [, dateColLetter, dateTopRow, , dateBotRow] = m;
  const firstDataRow = hasHeaders ? Number(dateTopRow) + 1 : Number(dateTopRow);
  const lastDataRow = Number(dateBotRow);
  const dataCount = Math.max(0, lastDataRow - firstDataRow + 1);
  if (dataCount === 0) {
    return { stepId: "", status: "success", message: "No data rows in dateColumn." };
  }

  const dateSheetName = dateUsed.worksheet.name;
  const sheetPrefix = dateSheetName.includes(" ") ? `'${dateSheetName}'!` : `${dateSheetName}!`;

  // Normalize output column — accept "G", "Sheet1!G", etc.
  let outColPart = outputColumn;
  let outSheet: string | null = null;
  if (outColPart.includes("!")) {
    const parts = outColPart.split("!");
    outSheet = parts[0].replace(/^'|'$/g, "");
    outColPart = parts[1];
  }
  outColPart = outColPart.replace(/[:$]/g, "").replace(/\d+$/, "");
  const outColLetter = outColPart.toUpperCase();
  if (!/^[A-Z]+$/.test(outColLetter)) {
    return { stepId: "", status: "error", message: `Invalid outputColumn: "${outputColumn}"` };
  }
  const outSheetName = outSheet ?? dateSheetName;
  const outSheetObj = context.workbook.worksheets.getItem(outSheetName);

  // Reference date: explicit or TODAY().
  const refDateExpr = referenceDate
    ? (() => {
        // Accept "dd/mm/yyyy" — convert to DATE(y,m,d) expression.
        const s = String(referenceDate).trim();
        const mm = s.match(/^(\d{1,2})\/(\d{1,2})\/(\d{2,4})$/);
        if (mm) {
          let y = Number(mm[3]);
          if (y < 100) y += 2000;
          return `DATE(${y},${Number(mm[2])},${Number(mm[1])})`;
        }
        const iso = s.match(/^(\d{4})-(\d{1,2})-(\d{1,2})$/);
        if (iso) return `DATE(${Number(iso[1])},${Number(iso[2])},${Number(iso[3])})`;
        return `DATEVALUE("${s.replace(/"/g, '""')}")`;
      })()
    : "TODAY()";

  // Sorted buckets ascending — labels are "0-B1", "B1+1-B2", ... "Bn+".
  const sortedBuckets = [...buckets].sort((a, b) => a - b);
  const bucketLabels: string[] = [];
  for (let i = 0; i < sortedBuckets.length; i++) {
    const lower = i === 0 ? 0 : sortedBuckets[i - 1] + 1;
    bucketLabels.push(`${lower}-${sortedBuckets[i]}`);
  }
  bucketLabels.push(`${sortedBuckets[sortedBuckets.length - 1]}+`);

  // Build formula. We reference the first data row of the date column
  // (relative) and auto-fill down.
  const firstDateCell = `${dateColLetter}${firstDataRow}`;
  const ageExpr = `(${refDateExpr}-${firstDateCell})`;
  const ifsParts: string[] = [];
  for (let i = 0; i < sortedBuckets.length; i++) {
    ifsParts.push(`${ageExpr}<=${sortedBuckets[i]},"${bucketLabels[i]}"`);
  }
  ifsParts.push(`TRUE,"${bucketLabels[bucketLabels.length - 1]}"`);
  const formula = `=IFS(${ifsParts.join(",")})`;

  try {
    // Optional header.
    if (hasHeaders) {
      outSheetObj.getRange(`${outColLetter}${Number(dateTopRow)}`).values = [["Age Bucket"]];
    }
    const anchor = outSheetObj.getRange(`${outColLetter}${firstDataRow}`);
    anchor.formulas = [[formula]];
    await context.sync();
    if (dataCount > 1) {
      const filldown = anchor.getResizedRange(dataCount - 1, 0);
      anchor.autoFill(filldown, Excel.AutoFillType.fillDefault);
      await context.sync();
    }
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    return { stepId: "", status: "error", message: `Failed to write aging column: ${msg}`, error: msg };
  }

  const outputAddr = `${sheetPrefix}${outColLetter}${firstDataRow}:${outColLetter}${lastDataRow}`;
  return {
    stepId: "",
    status: "success",
    message: `Wrote aging buckets (${bucketLabels.join(", ")}) to ${outputAddr}.`,
    outputs: { outputRange: outputAddr },
  };
}

registry.register(meta, handler as any);
export { meta };
